packages/build-utils/src/workspaces/get-workspace-package-paths.ts

```diff
@@ -0,0 +1,112 @@
+import _path from 'path';
+import yaml from 'js-yaml';
+import glob from 'glob';
+import { DetectorFilesystem } from '../detectors/filesystem';
+import { Workspace } from './get-workspaces';
+import { getGlobFs } from '../fs/get-glob-fs';
+
+const path = _path.posix;
+
+interface GetPackagePathOptions {
+  fs: DetectorFilesystem;
+}
+
+export interface GetWorkspacePackagePathsOptions extends GetPackagePathOptions {
+  fs: DetectorFilesystem;
+  workspace: Workspace;
+}
+
+export async function getWorkspacePackagePaths({
+  fs,
+  workspace,
+}: GetWorkspacePackagePathsOptions): Promise<string[]> {
+  const { type, rootPath: workspaceRootPath } = workspace;
+  const workspaceFs = fs.chdir(workspaceRootPath);
+
+  let results: string[] = [];
+
+  switch (type) {
+    case 'yarn':
+    case 'npm':
+      results = await getPackageJsonWorkspacePackagePaths({ fs: workspaceFs });
+      break;
+    case 'pnpm':
+      results = await getPnpmWorkspacePackagePaths({ fs: workspaceFs });
+      break;
+    default:
+      throw new Error(`Unknown workspace implementation: ${type}`);
+  }
+
+  return results.map(packagePath => {
+    return path.join(workspaceRootPath, path.dirname(packagePath));
+  });
+}
+
+type PackageJsonWithWorkspace = {
+  workspaces:
+    | {
+        packages: string[];
+        noHoist?: string[];
+      }
+    | string[];
+};
+
+type PnpmWorkspaces = {
+  packages: string[];
```

Comment:
```suggestion
  packages?: string[];
```

This should be optional for the case when its missing